ls1intum/Artemis
# [Feature] Extend tutor leaderboards with average_score and average_rating

### Checklist

[ ]* : not needed. 

### Motivation and Context


Tutors' Leaderboards are an important part of Artemis which show information about the performance of the tutors. 
There are 2 types of leaderboards: Course and Exercise leaderboards. 
The motivation for this PR is to **enrich the Tutor Leaderboards**, both for course and exercise, with **2 new columns** named **Average Score** and **Average Ratings**.  
The _average score_ indicates the average of all the scores that a tutor has given when reviewing students’ submission.
While the _average rating_ is the value that indicates how many stars in average(1-5) has a tutor received by the students.

### Description

I extended `TutorLeaderboardDTO` to contain `averageScore` and `averageRating` fields. 
`averageScore` and `averageRating` fields are computed dynamically by the database on the fly during the request. Extended the table to have two extra columns with some formatting of the double values. Added two integration tests that test the average computation on the server.

### Steps for Testing


1. Login to Artemis as an Instructor and create 1 exercise(any type). set the due date and assessment date with ~5 difference so you can have time to execute steps 2-3
2. Login to Artemis as (at least 2) different students, solve the exercise and submit.
3. Login to Artemis as a tutor and review both of the 2 student’s submissions. (Remember the tutor's name)
4. Login to Artemis with each student account and rate the tutor (this step is necessary to test the ratings)
5. Now, you can login to Artemis as a tutor/instructor and go to the Exercise Assessment Dashboard.
6. Scroll down until you reach the Tutor Exercise Leaderboard.
7. There should be 2 new columns: Average Score and Average Rating.
8. Verify with your calculator if the values displayed in the table are correct.
9. Don't forget to also check the values in the Tutor Course Leaderboard.

### Review Progress





- Code Review
- Manual Tests

### Test Coverage



TutorLeaderboardService.java: 100%

### Screenshots


<img width="1239" alt="Screenshot 2021-09-20 at 8 12 29 PM" src="https://user-images.githubusercontent.com/51077603/134053048-e3d2beff-2176-45c7-bdcd-eeb0ae7c6403.png">

- - -

<img width="1232" alt="Screenshot 2021-09-20 at 8 10 45 PM" src="https://user-images.githubusercontent.com/51077603/134053084-258c061f-e0a1-4408-b11d-470922b1bc9c.png">
